src/extlib/logger.hpp

```diff
@@ -0,0 +1,88 @@
+#ifndef CPPLOGGER_HPP
+#define CPPLOGGER_HPP
+
+#include <chrono>
+#include <iostream>
+#include <string>
+#include <fstream>
+#include <filesystem>
+#include <sstream>
+#include <iomanip>
+
+enum class LogLevel {
+    INFO,
+    WARNING,
+    ERROR,
+    DEBUG
+};
+
+enum class LogOutput {
+    CONSOLE,
+    FILE
+};
+
+class Logger {
+public:
+    static Logger* init_logging() {
+        static Logger instance;
+        return &instance;
+    }
+
+    void log(LogLevel log_level, LogOutput log_output,const std::string& message) {
+        std::string log_message = get_current_timestamp_str() +
+                                  get_log_level_str(log_level) +
+                                  message;
+        log_to_file(log_message);
+
+        if (log_output == LogOutput::CONSOLE) {
+            std::cout << log_message << std::endl;
+        }
+    }
+
+    void set_log_filepath(const std::string& filepath) {
+        _log_filepath = filepath;
+    }
+
+private:
+    std::string _log_filepath;
+
+    std::string get_log_level_str(LogLevel level) const {
+        switch (level) {
+            case LogLevel::ERROR:
+                return "[ERROR]  ";
+            case LogLevel::WARNING:
+                return "[WARN]   ";
+            case LogLevel::INFO:
+                return "[INFO]   ";
+            case LogLevel::DEBUG:
+                return "[DEBUG]  ";
+            default:
+                return "[UNKOWN] ";
```

Comment:
Let's not put default. The reason is if you don't put default, the compiler will generate a warning message saying that you did not cover all the enum values. This is great since the next time you add something to the enum, the compiler will remind you if you forget to update this function. If you use default, the compiler will not warn you. So it's better to not use default here.